src/Playroom/CodeEditor/keymaps/comment.ts

```diff
@@ -0,0 +1,329 @@
+      to = new Pos(to.line - 1);
+    }
+
+    const commentType = determineCommentType(cm, from, to);
+
+    const fullContent = cm.getRange(new Pos(from.line, 0), new Pos(to.line));
+    const existingIndent = fullContent.length - fullContent.trimStart().length;
+
+    const trimmedContent = fullContent.trim();
+
+    const isAlreadyCommented =
+      (trimmedContent.startsWith(BLOCK_COMMENT_START) &&
+        trimmedContent.endsWith(BLOCK_COMMENT_END)) ||
+      trimmedContent.startsWith(LINE_COMMENT_START);
+
+    const isMultiLineSelection = to.line !== from.line;
+
+    tagRanges.push({
+      from,
+      to,
+      multiLine: isMultiLineSelection,
+      existingIndent,
+      isAlreadyCommented,
+      commentType,
+    });
+
+    const fromOffset = getSelectionFromOffset({
+      commentType,
+      isAlreadyCommented,
+      fullContent,
+      from,
+    });
+
+    const toOffset = getSelectionToOffset({
+      to,
+      commentType,
+      isAlreadyCommented,
+      isMultiLineSelection,
+      fullContent,
+    });
+
+    const newSelectionRangeFrom = new Pos(from.line, from.ch + fromOffset);
+    const newSelectionRangeTo = new Pos(to.line, to.ch + toOffset);
+
+    const newSelection = isReverseSelection(range)
+      ? { anchor: newSelectionRangeTo, head: newSelectionRangeFrom }
+      : { anchor: newSelectionRangeFrom, head: newSelectionRangeTo };
+
+    newSelections.push(newSelection);
+  }
+
+  cm.operation(() => {
+    for (const range of [...tagRanges].reverse()) {
+      const newRangeFrom = new Pos(range.from.line, range.existingIndent);
+      const newRangeTo = new Pos(range.to.line);
+
+      const existingContent = cm.getRange(newRangeFrom, newRangeTo);
+
+      if (range.isAlreadyCommented) {
+        const uncommentType: CommentType = existingContent
+          .trimStart()
+          .startsWith(BLOCK_COMMENT_START)
+          ? 'block'
+          : 'line';
+
+        const existingContentWithoutComment = existingContent.replace(
+          uncommentType === 'block'
+            ? OPENING_AND_CLOSING_BLOCK_COMMENT_SYNTAX
+            : OPENING_LINE_COMMENT_SYNTAX_WITH_LEADING_WHITESPACE,
+          uncommentType === 'block' ? '' : LINE_COMMENT_LEADING_WHITESPACE
+        );
+        cm.replaceRange(
+          existingContentWithoutComment,
+          newRangeFrom,
+          newRangeTo
+        );
+      } else if (range.commentType === 'block') {
+        cm.replaceRange(
+          `${BLOCK_COMMENT_START} ${existingContent} ${BLOCK_COMMENT_END}`,
+          newRangeFrom,
+          newRangeTo
+        );
+      } else if (range.multiLine) {
+        const updatedContent = existingContent.replace(
+          /^(\s*)/gm,
+          `$1${LINE_COMMENT_START} `
+        );
+        cm.replaceRange(updatedContent, newRangeFrom, newRangeTo);
+      } else {
+        cm.replaceRange(
+          `${LINE_COMMENT_START} ${existingContent}`,
+          newRangeFrom,
+          newRangeTo
```

Comment:
Looks like these `replaceRange` calls only differ in there first parameter, so they can all probably be collapsed into a single call outside this `if else`.